kernel/ps2/src/lib.rs

```diff
@@ -1,484 +1,704 @@
+    port1_interrupt_enabled: bool1,
+    /// interrupt on [ControllerToHostStatus] `mouse_output_buffer_full`
+    port2_interrupt_enabled: bool1, //NOTE: only if 2 PS/2 ports supported
+    /// Cleared on reset; set when the system passed Power-on self-test
+    #[allow(dead_code)]
+    system_passed_self_test: bool1,
+    // or override_keyboard_inhibiting
+    #[allow(dead_code)]
+    should_be_zero: u1,
+    /// disables the keyboard
+    port1_clock_disabled: bool1,
+    /// disables the auxilary device (mouse)
+    port2_clock_disabled: bool1, //NOTE: only if 2 PS/2 ports supported
+    /// whether IBM scancode translation is enabled (0=AT, 1=PC)
+    #[allow(dead_code)]
+    port1_translation_enabled: bool1,
+    #[allow(dead_code)]
+    must_be_zero: u1,
 }
 
 /// read the config of the ps2 port
-pub fn ps2_read_config() -> u8 {
-    ps2_write_command(0x20);
-    let config = ps2_read_data();
-    config
+fn read_config() -> ControllerConfigurationByte {
+    write_command(ReadFromInternalRAMByte0);
+    ControllerConfigurationByte::from_bytes([read_data()])
 }
 
 /// write the new config to the ps2 command port (0x64)
-pub fn ps2_write_config(value: u8) {
-    ps2_write_command(0x60);
-    unsafe { PS2_PORT.lock().write(value) };
+fn write_config(value: ControllerConfigurationByte) {
+    write_command(WriteToInternalRAMByte0);
+    write_data(WritableData::Configuration(value));
 }
 
 /// initialize the first ps2 data port
 pub fn init_ps2_port1() {
-    //disable PS2 ports first
-    ps2_write_command(0xA7);
-    ps2_write_command(0xAD);
-
-    // clean the output buffer of ps2 to avoid conflicts
-    ps2_clean_buffer();
+    init_ps2_port(PS2Port::One);
+}
 
-    // set the configuration
-    {
-        // read the old configuration
-        let old_config = ps2_read_config();
+/// initialize the second ps2 data port
+pub fn init_ps2_port2() {
+    init_ps2_port(PS2Port::Two);
+}
+enum PS2Port {
+    One,
+    Two,
+}
 
-        // set the new configuration
-        let new_config = (old_config & 0xEF) | 0x01;
-        ps2_write_config(new_config);
+// see https://wiki.osdev.org/%228042%22_PS/2_Controller#Initialising_the_PS.2F2_Controller
+fn init_ps2_port(port: PS2Port) {
+    // Step 3: Disable Devices
+    write_command(DisablePort2);
+    write_command(DisablePort1);
+
+    // Step 4
+    flush_output_buffer();
+
+    // Step 5: Set the Controller Configuration Byte
+    let mut config = read_config();
+    match port {
+        PS2Port::One => {
+            config.set_port1_clock_disabled(false);
+            config.set_port1_interrupt_enabled(true);
+        }
+        PS2Port::Two => {
+            config.set_port2_clock_disabled(false);
+            config.set_port2_interrupt_enabled(true);
+        }
     }
+    write_config(config);
+
+    // Step 9: Enable Devices
+    write_command(EnablePort1);
+    match port {
+        PS2Port::Two => write_command(EnablePort2),
+        PS2Port::One => (),
+    };
```

Comment:
can you add a comment explaining why we unconditionally enable port 1 but only enable port 2 when initializing port 2? Seems a bit strange, and a future reader of this code may interpret it as an accidental oversight

Reply:
I just left it the way it was, because I'm not sure either, but will test.
(Just some rambling which might be useful later:
In wiki.osdev they initialize and test both keyboard and mouse at once, which we currently do separately, probably because of modularity.
I can just guess that we probably init the keyboard first, which means port 1 needs to be reenabled and then we init the mouse, so afterwards both need to be reenabled. This would mean we can't change the order of initialization anyways, because it would leave the mouse disabled.)

Reply:
oh that is odd. thanks for the clarification!

I think it's completely fine to initialize them together as part of a larger `ps2::init()` function or something like that, in which we follow your advice above and attempt to init keyboard, then mouse. That should be possible to directly invoke from the device manager or captain, wherever appropriate.